### PR TITLE
[CHK-2509] auth requested retry

### PR DESCRIPTION
**Pull request by ciuffagianluca**
Add Queue consumer for authorization requested retry events
Edit business logic to handle NPG getState invocation

#### List of Changes

- Added queue reference for retry event
- Added retry queue consumer
- Unify business logic for `auth requested` and `auth requested retry` event
- Implements new business logic for `getState` and `PATCH auth-request` responses handling. There are no cases that lead events in dead letter. If `getState` or `PATCH auth-request` response with 4xx error no retries are enqueued. If one of the responses is 5xx the retry are enqueued. The only cases that lead to write events in dead letter are the ones about event parsing failure or transaction aggregate parsing failure
- Added unit tests

#### Motivation and Context

Enforce NPG authorization process workflow

#### How Has This Been Tested?

Tests were performed by unit tests and in dev env

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -563,8 +563,8 @@
 				<configuration>
 					<skipCheckoutIfExists>false</skipCheckoutIfExists>
 					<connectionUrl>scm:git:https://github.com/pagopa/pagopa-ecommerce-commons.git</connectionUrl>
-					<scmVersionType>tag</scmVersionType>
-					<scmVersion>${pagopa-ecommerce-commons.version}</scmVersion>
+					<scmVersionType>branch</scmVersionType>
+					<scmVersion>CHK-2508-add-auth-requested-retry</scmVersion>
```
Comment (pietro-tota) [inline pom.xml:567]: remember to restore this params

File: helm/values-prod.yaml [hunk 2]
```diff
@@ -133,8 +136,8 @@ microservice-chart:
     CLOSE_PAYMENT_RETRY_EVENT_MAX_ATTEMPTS: "3"
     NOTIFICATION_RETRY_EVENT_BASE_INTERVAL_SECONDS: "600"
     NOTIFICATION_RETRY_EVENT_MAX_ATTEMPTS: "3"
-    AUTHORIZATION_REQUESTED_RETRY_EVENT_BASE_INTERVAL_SECONDS: "60"
-    AUTHORIZATION_REQUESTED_RETRY_EVENT_MAX_ATTEMPTS: "3"
+    AUTHORIZATION_OUTCOME_WAITING_EVENT_BASE_INTERVAL_SECONDS: "60"
```
Comment (pietro-tota) [inline helm/values-prod.yaml:139]: maybe this base interval value have to be lowered? what do you think?

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/TransactionAuthorizationRequestedQueueConsumer.kt [hunk 3]
```diff
@@ -1,85 +1,35 @@
...
 
-  var logger: Logger =
-    LoggerFactory.getLogger(TransactionAuthorizationRequestedQueueConsumer::class.java)
-
   fun messageReceiver(
-    parsedEvent: QueueEvent<TransactionAuthorizationRequestedEvent>,
+    parsedEvent:
+      Either<
+        QueueEvent<TransactionAuthorizationRequestedEvent>,
+        QueueEvent<TransactionAuthorizationOutcomeWaitingEvent>>,
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/TransactionAuthorizationRequestedQueueConsumer.kt:26]: Change this fun input to receive only `TransactionAuthorizationRequestedEvent` and then wrap it to Either left before pass it to the centralized logic. This way this handler can only receive the TransactionAuthorizationRequestedEvent

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/TransactionAuthorizationOutcomeWaitingQueueConsumer.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+@Service("TransactionAuthorizationOutcomeWaitingQueueConsumerV2")
+class TransactionAuthorizationOutcomeWaitingQueueConsumer(
+  @Autowired private val authorizationRequestedHelper: AuthorizationRequestedHelper,
+) {
+
+  fun messageReceiver(
+    parsedEvent:
+      Either<
+        QueueEvent<TransactionAuthorizationRequestedEvent>,
+        QueueEvent<TransactionAuthorizationOutcomeWaitingEvent>>,
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/TransactionAuthorizationOutcomeWaitingQueueConsumer.kt:22]: same as below comment, pass here only `TransactionAuthorizationOutcomeWaitingEvent` and wrap to Either.right below

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/helpers/AuthorizationRequestedHelper.kt [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+import it.pagopa.ecommerce.eventdispatcher.utils.DeadLetterTracedQueueAsyncClient
+import java.util.*
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.switchIfEmpty
+
+data class AuthorizationRequestedEvent(
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/helpers/AuthorizationRequestedHelper.kt:27]: in this particular case an Either should be sufficient, this data class can be useful when more than 2 variants are required as input.

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/helpers/AuthorizationRequestedHelper.kt [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+  @Autowired private val authorizationStateRetrieverService: AuthorizationStateRetrieverService,
+  @Autowired private val deadLetterTracedQueueAsyncClient: DeadLetterTracedQueueAsyncClient,
+  @Autowired private val tracingUtils: TracingUtils,
+  @Autowired private val strictSerializerProviderV2: StrictJsonSerializerProvider
+) {
+
+  var logger: Logger = LoggerFactory.getLogger(AuthorizationRequestedHelper::class.java)
+
+  fun authorizationStateRetrieve(
+    parsedEvent: AuthorizationRequestedEvent,
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/helpers/AuthorizationRequestedHelper.kt:82]: see above comment, here can be used Either data structure: is well known and more readable than a custom class (and in this situation there are only two variant)

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/helpers/AuthorizationRequestedHelper.kt [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+    val authorizationRequestedPipeline =
+      transactionsEventStoreRepository
+        .findByTransactionIdOrderByCreationDateAsc(transactionId)
+        .reduce(EmptyTransaction(), Transaction::applyEvent)
+        .cast(BaseTransaction::class.java)
+        .filter { it.status == TransactionStatusDto.AUTHORIZATION_REQUESTED }
+        .switchIfEmpty {
+          logger.info(
+            "Transaction $transactionId is not is Authorization Requested status. No more action needed")
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/queues/v2/helpers/AuthorizationRequestedHelper.kt:97]:
```suggestion
            "Transaction [$transactionId] status: [${it.status}]. No action needed")
```

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/v2/AuthorizationStateRetrieverService.kt [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import it.pagopa.ecommerce.commons.utils.NpgPspApiKeysConfig
+import it.pagopa.ecommerce.eventdispatcher.exceptions.NpgBadRequestException
+import it.pagopa.ecommerce.eventdispatcher.exceptions.NpgServerErrorException
+import java.util.*
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Service
+import reactor.core.publisher.Mono
+
+@Service
+class AuthorizationStateRetrieverService(
```
Comment (pietro-tota) [inline src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/v2/AuthorizationStateRetrieverService.kt:16]: is this class a duplicate of this [one](https://github.com/pagopa/pagopa-ecommerce-event-dispatcher-service/blob/main/src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/v2/NpgStateService.kt)?
Code seems to be duplicated

File: README.md [hunk 9]
```diff
@@ -83,8 +84,8 @@ These are all environment variables needed by the application:
...
-| AUTHORIZATION_REQUESTED_RETRY_EVENT_MAX_ATTEMPTS             | Max attempts to be performed for payment gateway auth state request                                                                                                                                                                                                | string  |               |
+| AUTHORIZATION_OUTCOME_WAITING_EVENT_BASE_INTERVAL_SECONDS    | Base interval used to calculate visibility for next payment gateway auth state request event                                                                                                                                                                       | string  |               |
+| AUTHORIZATION_OUTCOME_WAITING_EVENT_MAX_ATTEMPTS             | Max attempts to be performed for payment gateway auth state request                                                                                                                                                                                                | string  |               |
```
Comment (pietro-tota) [inline README.md:88]: add missing `AUTHORIZATION_OUTCOME_WAITING_EVENT_PAYMENT_TOKEN_VALIDITY_TIME_OFFSET`

File: src/main/resources/application.properties [hunk 10]
```diff
@@ -48,9 +49,9 @@ closePaymentRetry.eventOffsetSeconds=${CLOSE_PAYMENT_RETRY_EVENT_BASE_INTERVAL_S
 closePaymentRetry.maxAttempts=${CLOSE_PAYMENT_RETRY_EVENT_MAX_ATTEMPTS}
 closePaymentRetry.paymentTokenValidityTimeOffset=${CLOSE_PAYMENT_RETRY_EVENT_PAYMENT_TOKEN_VALIDITY_TIME_OFFSET:10}
 
-transactionAuthorizationRequested.eventOffsetSeconds=${AUTHORIZATION_REQUESTED_RETRY_EVENT_BASE_INTERVAL_SECONDS}
-transactionAuthorizationRequested.maxAttempts=${AUTHORIZATION_REQUESTED_RETRY_EVENT_MAX_ATTEMPTS}
-transactionAuthorizationRequested.paymentTokenValidityTimeOffset=${AUTHORIZATION_REQUESTED_RETRY_EVENT_PAYMENT_TOKEN_VALIDITY_TIME_OFFSET:10}
+transactionAuthorizationOutcomeWaiting.eventOffsetSeconds=${AUTHORIZATION_OUTCOME_WAITING_EVENT_BASE_INTERVAL_SECONDS}
+transactionAuthorizationOutcomeWaiting.maxAttempts=${AUTHORIZATION_OUTCOME_WAITING_EVENT_MAX_ATTEMPTS}
+transactionAuthorizationOutcomeWaiting.paymentTokenValidityTimeOffset=${AUTHORIZATION_OUTCOME_WAITING_EVENT_PAYMENT_TOKEN_VALIDITY_TIME_OFFSET:10}
```
Comment (pietro-tota) [inline src/main/resources/application.properties:54]: can we explicit time unit for this time offset variable as done for `AUTHORIZATION_OUTCOME_WAITING_EVENT_BASE_INTERVAL_SECONDS`

**Review by pietro-tota — APPROVED** [review]
🚀 well done, just some trivial issues

File: src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/v2/AuthorizationStateRetrieverService.kt (hunk 8 above)
Comment (ciuffagianluca) [reply src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/v2/AuthorizationStateRetrieverService.kt:16]: That's right. A rebase from main did resurrect it

**Review by pietro-tota — DISMISSED** [review]

File: helm/values-prod.yaml (hunk 2 above)
Comment (pietro-tota) [reply helm/values-prod.yaml:139]: Remain only this parameter value to be discussed as 1 minute retry offset maybe is too large given the actual f.e. polling timeout

**Review by pietro-tota — APPROVED** [review]
🚀

File: helm/values-prod.yaml (hunk 2 above)
Comment (ciuffagianluca) [reply helm/values-prod.yaml]: Se to 10. Passible of future evaluation.

**Review by pietro-tota — APPROVED** [review]

**Pull request merged**